### PR TITLE
Corrected traceMult

### PR DESCRIPTION
**Pull request by linneamk**
This change is [Reviewable](https://reviewable.io/reviews/lanl/bml/313)

### CODE REVIEW

**Comment by nicolasbock** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please start subject with capital letter.<br />
8511136b4c1436ab671aa5f68867af904b3f0250</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: src/C-interface/bml_trace.c [hunk 1]
```diff
@@ -7,7 +7,7 @@
 #include "ellblock/bml_trace_ellblock.h"
 
 #include <stdlib.h>
-
+#include <stdio.h>
```
Comment (nicolasbock) [inline src/C-interface/bml_trace.c]: Why do we need stdio.h?

File: src/C-interface/ellpack/bml_trace_ellpack.c [hunk 2]
```diff
@@ -5,6 +5,7 @@
 #include "bml_types_ellpack.h"
 
 #include <stdlib.h>
+#include <stdio.h>
```
Comment (nicolasbock) [inline src/C-interface/ellpack/bml_trace_ellpack.c]: Same here, why add studio.h?

File: src/C-interface/ellpack/bml_trace_ellpack_typed.c [hunk 3]
```diff
@@ -104,6 +104,7 @@ double TYPED_FUNC(
     //for (int i = 0; i < A_N; i++)
     for (int i = A_localRowMin[myRank]; i < A_localRowMax[myRank]; i++)
     {
+
```
Comment (nicolasbock) [inline src/C-interface/ellpack/bml_trace_ellpack_typed.c]: Unnecessary empty line

File: src/C-interface/bml_trace.c (hunk 1 above)
Comment (linneamk) [reply src/C-interface/bml_trace.c]: We don't, I removed it.

File: tests/C-tests/test_trace_mult_typed.c [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    double trace, trace_mult, diff, threshold;
+    threshold = 0.0;
+    REAL_T scalar = 0.5;
+
+    A = bml_identity_matrix(matrix_type, matrix_precision, N, M, sequential);
+    B = bml_scale_new(&scalar, A);
+    bml_scale_inplace(&scalar, A);
+
+    trace_mult = bml_trace_mult(A, B);
+    trace = scalar * scalar * N;
```
Comment (nicolasbock) [inline tests/C-tests/test_trace_mult_typed.c]: Thanks for adding this test! @cnegre do you think the test is sufficient as it is written? It's testing only a^2 Id and not a more general case.

**Comment by cnegre** [comment]
Here we need to make sure we test the general case. The "true" trace will be the one obtained by explicitly computing Sum_ik A_ik B_ki in dense format.

File: tests/C-tests/test_trace_mult_typed.c [hunk 5]
```diff
@@ -23,14 +23,28 @@ int TYPED_FUNC(
...
+
+    trace = 0;
+    for (int i = 0; i < N; i++)
+    {
+        for (int k = 0; k < N; k++)
+        {
+            trace +=
+                (*(REAL_T *) bml_get(A, i, k)) *
+                (*(REAL_T *) bml_get(B, k, i));
+        }
```
Comment (nicolasbock) [inline tests/C-tests/test_trace_mult_typed.c]: Something like this @cnegre ?

**Comment by nicolasbock** [comment]
The test should work now (I hope). The current implementation of trace_mult assumes symmetric matrices.

**Comment by jeanlucf22** [comment]
Looks like tolerance is too tight for single precision tests.

**Comment by linneamk** [comment]
I did something like this too. It works if you divide the error by N*N.

**Comment by nicolasbock** [comment]
Thanks @jeanlucf22 and @linneamk ! I changed the comparison to `tol/N^2`.

File: tests/C-tests/test_trace_mult_typed.c (hunk 5 above)
Comment (cnegre) [reply tests/C-tests/test_trace_mult_typed.c]: Perfect!

**Comment by nicolasbock** [comment]
Thanks! We can merge then once CI is done.

**Pull request merged**